x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go

```diff
@@ -159,67 +164,70 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) error {
 				continue
 			}
 
-			if namespaceResourceType.metricNames != nil && namespaceResourceType.dimensions == nil {
-				for _, listMetric := range listMetricsOutput {
-					if exists, _ := aws.StringInSlice(*listMetric.MetricName, namespaceResourceType.metricNames); !exists {
-						continue
-					}
-					filteredMetricWithStatsTotal = append(filteredMetricWithStatsTotal,
-						metricsWithStatistics{
-							cloudwatchMetric: listMetric,
-							statistic:        namespaceResourceType.statistic,
-						})
-
-					exists, _ := aws.StringInSlice(namespaceResourceType.resourceTypeFilter, resourceTypes)
-					if namespaceResourceType.resourceTypeFilter != "" && !exists {
-						resourceTypes = append(resourceTypes, namespaceResourceType.resourceTypeFilter)
+			for _, metricD := range metricDetails {
+				if metricD.resourceTypeFilter != "" {
+					if _, ok := resourceTypes[metricD.resourceTypeFilter]; ok {
+						resourceTypes[metricD.resourceTypeFilter] = append(resourceTypes[metricD.resourceTypeFilter], metricD.tags...)
+					} else {
+						resourceTypes[metricD.resourceTypeFilter] = metricD.tags
 					}
 				}
-			} else if namespaceResourceType.metricNames == nil && namespaceResourceType.dimensions != nil {
-				for _, listMetric := range listMetricsOutput {
-					if !compareAWSDimensions(listMetric.Dimensions, namespaceResourceType.dimensions) {
-						continue
-					}
-					filteredMetricWithStatsTotal = append(filteredMetricWithStatsTotal,
-						metricsWithStatistics{
-							cloudwatchMetric: listMetric,
-							statistic:        namespaceResourceType.statistic,
-						})
-
-					exists, _ := aws.StringInSlice(namespaceResourceType.resourceTypeFilter, resourceTypes)
-					if namespaceResourceType.resourceTypeFilter != "" && !exists {
-						resourceTypes = append(resourceTypes, namespaceResourceType.resourceTypeFilter)
+			}
+
+			var filteredMetricWithStatsTotal []metricsWithStatistics
+			for _, listMetric := range listMetricsOutput {
```

Comment:
could you please add a comment on these loops to explain what each of them is doing?

Reply:
Yes, I'm actually gonna pull these loops into their own functions and add some tests as well 😬 Current code is too hard to read...